STM32/AC/Core/Src/ACBoard.c

```diff
@@ -188,63 +224,113 @@ static void CAallOn(bool isOn)
 }
 static void CAportState(int port, bool state, int percent, int duration)
 {
-	uint8_t pwmPercent = getPWMPinPercent(port-1);
-	// If heat sink has reached the maximum allowed temperature and user
-	// tries to heat the system further up then disregard the input command
-	if (heatSinkTemperature > MAX_TEMPERATURE && percent > pwmPercent)
-		return;
+    uint8_t pwmPercent = getPWMPinPercent(port-1);
+    // If heat sink has reached the maximum allowed temperature and user
+    // tries to heat the system further up then disregard the input command
+    if (heatSinkTemperature > MAX_TEMPERATURE && percent > pwmPercent)
+        return;
 
     actuatePins((ActuationInfo) { port - 1, percent, 1000*duration, true });
 }
 
+/*!
+** @brief Loop for controlling board temperature.
+**
+** Loops keeping the board cool using the attached fan. If the board gets too hot, starts to reduce
+** the power on all ports by reducing the PWM
+*/
 static void heatSinkLoop()
 {
-	static unsigned long previous = 0;
+    static unsigned long previous = 0;
     // Turn on fan if temp > 55 and turn of when temp < 50.
-    if (heatSinkTemperature < 50 && !isFanForceOn)
+    if (heatSinkTemperature <= MAX_TEMPERATURE)
     {
-		stmSetGpio(fanCtrl, false);
+        if (heatSinkTemperature < 50 && !isFanForceOn)
+        {
+            stmSetGpio(fanCtrl, false);
+        }
+        else if (heatSinkTemperature > 55)
+        {
+            stmSetGpio(fanCtrl, true);
+        }
+
+        bsClearField(BS_OVER_TEMPERATURE_Msk);
     }
-    else if (heatSinkTemperature > 55)
+    else // Safety mode to avoid overheating of the board
     {
-    	stmSetGpio(fanCtrl, true);
+        stmSetGpio(fanCtrl, true);
+        
+        unsigned long now = HAL_GetTick();
+        if (now - previous > 2000)  // 2000ms. Should be aligned with the control period of heater.
+        {
+            previous = now;
+            adjustPWMDown();
+        }
+
+        bsSetError(BS_OVER_TEMPERATURE_Msk);
     }
-    else if (heatSinkTemperature > MAX_TEMPERATURE) // Safety mode to avoid overheating of the board
+
+    setBoardTemp(heatSinkTemperature);
+}
+
+/*!
+** @brief Updates the global error/status object.
+**
+** Adds the port and fan status bits into the error field, and clears the error bit if there are no
+** more errors.
+*/
+static void updateErrorField() 
```

Comment:
Maybe a better name would be "updateBoardStatus" since it also relates to states that are not error related.